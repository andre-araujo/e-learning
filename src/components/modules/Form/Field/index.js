import { Component } from 'react';
import {
  func,
  string,
  shape,
  object,
} from 'prop-types';

class Field extends Component {
  componentDidMount() {
    const {
      name,
      initialValue,
      context,
    } = this.props;

    if (!context.fields[name]) {
      context.setFieldValue(name, initialValue);
    }
  }

  render() {
    const {
      render,
      name,
      initialValue = '',
      context,
      ...props
    } = this.props;

    let value = context.fields[name];

    if (value === undefined) {
      value = initialValue;
    }

    return render({
      ...props,
      name,
      value,
      onChange: e => context.setFieldValue(name, e.target.value),
    });
  }
}

Field.defaultProps = {
  initialValue: '',
};

Field.propTypes = {
  render: func.isRequired,
  name: string.isRequired,
  initialValue: string,
  context: shape({
    fields: object.isRequired,
    setFieldValue: func.isRequired,
  }).isRequired,
};

export default Field;
