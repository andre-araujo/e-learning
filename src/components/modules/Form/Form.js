import React, { createContext, Component } from 'react';
import { node, func, string } from 'prop-types';
import Field from './Field';

export const FormContext = createContext();

class Form extends Component {
  static Field = props => (
    <FormContext.Consumer>
      {
        context => <Field context={context} {...props} />
      }
    </FormContext.Consumer>
  )

  state = {
    fields: {},
  }

  setFieldValue = (name, value) => {
    this.setState(({ fields }) => ({
      fields: {
        ...fields,
        [name]: value,
      },
    }));
  }

  handleSubmit = (e) => {
    const { fields } = this.state;
    const { onSubmit } = this.props;

    e.preventDefault();
    onSubmit(fields);
  }

  render() {
    const {
      children,
      name,
      onSubmit,
      ...props
    } = this.props;
    const { fields } = this.state;

    const providerValue = {
      fields,
      setFieldValue: this.setFieldValue,
    };

    return (
      <form onSubmit={this.handleSubmit} name={name} {...props}>
        <FormContext.Provider value={providerValue}>
          {children}
        </FormContext.Provider>
      </form>
    );
  }
}

Form.defaultProps = {
  children: null,
};

Form.propTypes = {
  children: node,
  onSubmit: func.isRequired,
  name: string.isRequired,
};

export default Form;
