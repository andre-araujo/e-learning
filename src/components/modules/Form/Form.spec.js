import React from 'react';
import { mount } from 'enzyme';

import Form from './Form';

describe('Form component', () => {
  it('should match snapshot', () => {
    expect(mount(<Form />)).toMatchSnapshot();
  });
});
