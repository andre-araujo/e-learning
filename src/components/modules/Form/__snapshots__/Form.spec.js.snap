// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form component should match snapshot 1`] = `
<form
  name="form"
  onSubmit={[Function]}
>
  <ContextProvider
    value={
      Object {
        "fields": Object {},
        "setFieldValue": [Function],
      }
    }
  />
</form>
`;
